src/ucp/core/ucp_mm.c

```diff
@@ -1305,3 +1406,226 @@ ucs_status_t ucp_mem_reg_md_map_update(ucp_context_h context)
+                ucp_mem_h *memh_p)
+{
+    ucs_rcache_t *rcache = NULL;
+    ucp_mem_h memh, rcache_memh;
+    ucs_status_t status;
+    ucp_unpacked_exported_memh_t unpacked_memh;
+    ucs_rcache_region_t *rregion;
+    khiter_t iter;
+
+    status = ucp_memh_exported_unpack(context, export_mkey_buffer,
+                                      &unpacked_memh);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucp_memh_create(context, unpacked_memh.address,
+                             unpacked_memh.length, unpacked_memh.mem_type,
+                             UCT_ALLOC_METHOD_LAST, UCP_MEMH_FLAG_IMPORTED,
+                             &memh);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    if (ucs_likely(context->imported_mem_hash != NULL)) {
+        /* Try to find rcache of imported memory buffers for the specific peer
+         * with UUID packed in the exported_mkey_buffer */
+        UCP_THREAD_CS_ENTER(&context->mt_lock);
+
+        iter = kh_get(ucp_context_imported_mem_hash,
+                      context->imported_mem_hash, unpacked_memh.remote_uuid);
+        if (iter != kh_end(context->imported_mem_hash)) {
+            /* Found rcache for the specific peer */
+            rcache = kh_value(context->imported_mem_hash, iter);
+            ucs_assert(rcache != NULL);
+
+            rregion = ucs_rcache_lookup_unsafe(rcache, unpacked_memh.address,
+                                               unpacked_memh.length,
+                                               PROT_READ | PROT_WRITE);
+            if (rregion != NULL) {
+                rcache_memh = ucs_derived_of(rregion, ucp_mem_t);
+                if (ucs_likely(rcache_memh->reg_id == unpacked_memh.reg_id)) {
+                    ucp_memh_rcache_print(rcache_memh, unpacked_memh.address,
+                                          unpacked_memh.length);
+                    memh->parent = rcache_memh;
+                    status       = UCS_OK;
+                    UCP_THREAD_CS_EXIT(&context->mt_lock);
+                    goto out_memh_update;
+                }
+
+                ucs_debug("found memh %p (reg_id %" PRIu64 ") in rcache %p,"
+                          " but reg_id is not matched with %" PRIu64,
+                          rcache_memh, rcache_memh->reg_id, rcache,
+                          unpacked_memh.reg_id);
+
+                /* If registration IDs are not matched, but a region still
+                 * exists in the RCACHE of imported regions, it means that
+                 * an exported memory handle has already been destroyed for a
+                 * given address, but an imported memory handle hasn't been
+                 * retrieved from the RCACHE yet. So, it should have reference
+                 * counter == 1, since ucp_mem_unmap() should be invoked for
+                 * unused imported memory handles and then - for corresponding
+                 * exported ones */
```

Comment:
somehow my previous comment disappeared. Please rephrase it somehow, as this limitation is supposed to be fixed in fw

Reply:
but I don't mention the issue in FW.
this comments asks to support flow when imported keys are destroyed before corresponding exported keys are destroyed

Reply:
yes, this is the limitation which is not really needed.
can ignore for now, it's not an APi anyway

Reply:
ok